docs/design/branch-mode.md

```diff
@@ -0,0 +1,117 @@
+# Branch mode
+
+This is a proposed configuration option to replace
+`experimental-advance-branches`. It is not intended to replace or obviate
+topics.
+
+The design below assumes that the configuration option is enabled. There should
+be no change from current behavior if the option is not enabled.
+
+**"Non-headless"** or **"branch"** mode: a git-colocated repository with the
+new configuration option enabled is considered to be in this mode whenever
+`.git/HEAD` (hereafter `HEAD`) is populated with a valid `jj` bookmark, and the
+bookmark is currently at `@` or `@-`. In this design, the bookmark will be
+referred to as `b(HEAD)`.
+
+**Note:** The only part of this design that is specific to `git` colocation is
+the use of `.git/HEAD`. The design could be implemented for any backend that
+has a concept of a "current" branch.
+
+## Invariants
+
+The configuration option will never affect any of these `jj` behaviors:
+
+* how `@` is updated by each `jj` command
+* the way commit topology or contents (ignoring bookmarks) evolve with each
+  command
+
+## Diagrams
+
+These Excalidraw diagrams show the effect of various `jj` commands when in
+branch mode:
+https://excalidraw.com/#json=pkmTWTuzkToGdc4xSUy2C,7uhU9FBI6rJK-U_XGCH4HA
```

Comment:
Possibly adding on to https://github.com/martinvonz/jj/pull/4670#discussion_r1807534569 (which is the kind of questions these diagrams are probably meant to address), I found these diagrams hard to follow. They are very pretty, but they need more words telling the reader where to look, and what is most important about the pictures.

Another (more minor, I mostly figured it out but it could be clearer) thing I found confusing at a glance is what the red and the blue arrows mean. After looking at the text, they are meant to be bookmarks (you should label that in the picture!), but it's unclear how they are related to each other. I guess the red one is supposed to be the "current bookmark"?

I think the pictures could be very helpful once you make thing clearer. (Though if presenting a few clear examples as Martin asked is easier in other ways, that's more important). Here's one possible way to arrange them:

```
- Starting state
- A bit of text introducing a few examples
- A few examples (the pictures)
- A bit of text introducing a few more examples
- A few more examples
```

and so on.


----------

Another suggestion for the pictures: it would be interesting what happens if there are two bookmarks pointing to `Z`, only one of which is "current". You don't have to show it in every picture, though, or perhaps it could be in a separate picture.

Reply:
BTW, are the "split" and "new" diagrams switched?

Reply:
In your "ambiguous update: `abandon Z` example, I thing `jj` would currently create a branch conflict with the red branch being conflicted between pointing at D and at Y. The thing you suggested wouldn't work if we continue to disallow HEAD pointing to @ (as we currently do).

Reply:
Understood; I intentionally left the diagrams in Excalidraw rather than rendering SVGs so they could be iterated on. I'll certainly add labels and/or a legend to the diagram itself.

The basic jist is that the top left diagram is an "initial state", and then almost all the other diagrams show what the next state would be after executing the command shown from that initial state.

The branch arrows are unrelated; red happens to be the "current" branch in the initial state. The thing showing that it's "current", though, is the shading and coloring on the pointed-to commit.

Agreed the two-bookmarks example would be good to show; that's actually one of the primary things I struggle with using the current `advance-branches` feature.

Reply:
No, `split` and `new` are not switched, but the use of `new` as a node label is probably not helpful. `split` is similar to `new` in that both introduce a new commit on top of the target commit. (The docs even mention that if the target commit is empty, `split` is prohibited because it would be identical to `new`.)

One thing that may not be clear is that the letters/words/symbols in the commit nodes are change IDs, and `@` itself is shown as a commit. The command is always shown on the upper-left of each diagram. So there are two variants of `split` next to each other, one with no arguments and one with the argument `@-`. The commit-node that says `new` is just to indicate that a new commit has been created, but looking again, there's no good reason to use `Z'` in one `split` diagram and `new` in the other. I think the thing I need to do with those is explicitly state that the _content_ of the original `Z` commit has been split between `Z'` and `Z` (on the left) and between `Z` and `new` (on the right).

Reply:
> No, split and new are not switched,

It's then confusing to me how in the `new` example, `N` becomes a child of "new". Perhaps that's a mistake? If it is intentional, what's the logic behind it?

I think I was indeed mistaken about "split". It was partially because it is hard for me to not think of the highlighted commit as the working copy, it looks more important than the commit actually marked `@`, so I find myself getting them confused.

Reply:
Oh, yes, you're right, I think the `new` diagram is indeed incorrectly copy-pasted from `split @-`. Thank you. The `N` should still be a child of `Z`, as in `split` without args.

...speaking of args, I also forgot that the positional args of `split` are paths, not refs. So that should be `split -r @-` and `split` (without `-r`).

Reply:
Here are the updated diagrams: https://excalidraw.com/#json=SN8HJ4EXv3ChyzvUdra_i,lwYuJ7ddFRR5zMmGCDdI8w